hypervisor/src/kvm.rs

```diff
@@ -0,0 +1,1269 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright © 2020, Microsoft  Corporation
+//
+// Copyright 2018-2019 CrowdStrike, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use crate::cpu;
+use crate::hv;
+use crate::vm;
+
+#[cfg(target_arch = "aarch64")]
+use crate::aarch64::check_required_kvm_extensions;
+use crate::common::{CreateDevice, DeviceFd, IoEventAddress, IrqRouting, MemoryRegion};
+use crate::common::{FpuState, MpState, SpecialRegisters, StandardRegisters, VcpuEvents, VcpuExit};
+#[cfg(target_arch = "x86_64")]
+use crate::x86_64::check_required_kvm_extensions;
+#[cfg(target_arch = "x86_64")]
+use crate::x86_64::{CpuId, ExtendedControlRegisters, LapicState, MsrEntries, Xsave};
+#[cfg(target_arch = "x86_64")]
+use devices::ioapic;
+#[cfg(target_arch = "x86_64")]
+use kvm_bindings::{kvm_enable_cap, KVM_CAP_SPLIT_IRQCHIP};
+use kvm_ioctls::{Cap, Kvm, NoDatamatch, VcpuFd, VmFd};
+use std::result;
+use std::sync::Arc;
+use vm_memory::{Address, GuestAddress};
+
+use vmm_sys_util::eventfd::EventFd;
+pub const KVM_TSS_ADDRESS: GuestAddress = GuestAddress(0xfffb_d000);
+
+/// Wrapper over KVM VM ioctls.
+pub struct KvmVm {
+    fd: Arc<VmFd>,
+}
+///
+/// Implementation of Vm trait for KVM
+/// Example:
+/// hv = KvmHyperVisor::new().unwrap()
+/// vm = hv.create_vm().unwrap()
+/// vm.create_irq_chip().map_err()
+///
+impl vm::Vm for KvmVm {
+    #[cfg(target_arch = "x86_64")]
+    /// Sets the address of the three-page region in the VM's address space.
+    ///
+    /// See the documentation for `KVM_SET_TSS_ADDR`.
+    ///
+    /// # Arguments
+    ///
+    /// * `offset` - Physical address of a three-page region in the guest's physical address space.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # extern crate hypervisor;
+    /// # use hypervisor::KvmHyperVisor;
+    /// use hypervisor::KvmVm;
+    /// let hv = KvmHyperVisor::new().unwrap();
+    /// let vm = hv.create_vm().unwrap();
+    /// vm.set_tss_address(0xfffb_d000).unwrap();
+    /// ```
+    ///
+    fn set_tss_address(&self, offset: usize) -> vm::Result<()> {
+        self.fd
+            .set_tss_address(offset)
+            .map_err(|e| vm::HypervisorVmError::SetTssAddress(e.into()))
```

Comment:
I think you can simplify every
```rust
map_err(|e| vm::HypervisorVmError::SpecificError(e.into()))
```
with
```rust
map_err(vm::HypervisorVmError::SpecificError)
```

Reply:
If I change this I get the following error. At first I was trying to keep it kvm_ioctls::errorno::Error but in order to make it hypervisor agnostic Samuel suggested to use the current implementation.
`SetTssAddress(#[source] anyhow::Error),
TSS address error

type mismatch in function arguments
expected signature of `fn(vmm_sys_util::errno::Error) -> _`rustc(E0631)
vm.rs(31, 5): found signature of `fn(anyhow::Error) -> _`

Reply:
Oh I see. You need to convert the `kvm_ioctls::errorno::Error` error that you get from `kvm-ioctls` into `anyhow::Error`, so that from the trait we specify `anyhow::Error` only.